DESIGN.md

```diff
@@ -0,0 +1,370 @@
+# Apollo Android design discussion
+
+These are some preliminary ideas for mapping GraphQL query results to Java, based on what we currently do for Swift in Apollo iOS.
+
+## Query classes
+
+I would expect us to generate classes for queries (and other operations). Instances of these classes represent a particular query, fully configured with the required variables.
+
+These query objects can be used multiple times to parse a response or to load results from the cache. They can also be stored by a query manager that watches the store and reloads results when the contents change for example.
+
+We'll have to figure out what makes sense for Java, but one option seems to be to specify a generic type parameter on the `GraphQLQuery` interface to refer to the specific `Data` type that represents results for this query. (I've used `Data` here because this corresponds to the `data` field in a GraphQL response.)
+
+So something like this:
+
+```graphql
+query HeroAndFriendsNames($episode: Episode) {
+  hero(episode: $episode) {
+    name
+    friends {
+      name
+    }
+  }
+}
+```
+
+```java
+interface GraphQLQuery<D> {
+  D parseResponse(...);
+  // etc.
+}
+
+enum Episode {
+    NEWHOPE,
+    EMPIRE,
+    JEDI
+}
+
+class HeroAndFriendsNameQuery implements GraphQLQuery<HeroAndFriendsNameQuery.Data> {
+    private Episode episode;
+
+    public HeroAndFriendsNameQuery(Episode episode) {
+        this.episode = episode;
+    }
+
+    interface Data {
+        Hero hero();
+
+        interface Hero {
+            String name();
+            List<Friend> friends();
+
+            interface Friend {
+                String name();
+            }
+        }
+    }
+}
+```
+
+Nested interfaces are named based on the response name, not the field type or field name. Response names are guaranteed to be unique within a selection set, and it also seems to make more semantic sense. (A field `author` would refer to `Author` for example, and not to `User`.)
```

Comment:
I'd rather only add a prefix if we really have to (for deduping) in order to avoid clutter.